<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mendelian randomization • gwasglue</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mendelian randomization">
<meta property="og:description" content="gwasglue">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">gwasglue</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cojo.html">Conditional analysis of VCF files</a>
    </li>
    <li>
      <a href="../articles/colocalisation.html">Genetic colocalisation</a>
    </li>
    <li>
      <a href="../articles/finemapping.html">Clumping and finemapping</a>
    </li>
    <li>
      <a href="../articles/gwas2020.html">Major changes to the IEU GWAS resources for 2020</a>
    </li>
    <li>
      <a href="../articles/mr.html">Mendelian randomization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrcieu/gwasglue/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mendelian randomization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mrcieu/gwasglue/blob/master/vignettes/mr.Rmd"><code>vignettes/mr.Rmd</code></a></small>
      <div class="hidden name"><code>mr.Rmd</code></div>

    </div>

    
    
<p>To use the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> for MR analysis, see the <a href="https://mrcieu.github.io/TwoSampleMR/">TwoSampleMR</a> R package.</p>
<p>Here we’ll demonstrate how to achieve the same data extractions using the <a href="https://github.com/MRCIEU/gwas_vcf_spec">GWAS VCF</a> files.</p>
<p>We’ll use the example of LDL cholesterol <a href="https://gwas.mrcieu.ac.uk/datasets/ieu-a-300/">ieu-a-300</a> and coronary heart disease <a href="https://gwas.mrcieu.ac.uk/datasets/ieu-a-7/">ieu-a-7</a>.</p>
<p>Load libraries:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span>({
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TwoSampleMR</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">gwasglue</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">gwasvcf</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ieugwasr</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
}))
<span class="co">#&gt; API: public: http://gwas-api.mrcieu.ac.uk/</span></pre></body></html></div>
<div id="using-twosamplemr" class="section level2">
<h2 class="hasAnchor">
<a href="#using-twosamplemr" class="anchor"></a>Using TwoSampleMR</h2>
<p>This is a simple procedure for MR using the TwoSampleMR package:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Extract the instruments for LDL</span>
<span class="no">expd1</span> <span class="kw">&lt;-</span> <span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/extract_instruments.html">extract_instruments</a></span>(<span class="st">"ieu-a-300"</span>)

<span class="co"># Extract those SNP effects for CHD</span>
<span class="no">outd1</span> <span class="kw">&lt;-</span> <span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/extract_outcome_data.html">extract_outcome_data</a></span>(<span class="no">expd1</span>$<span class="no">SNP</span>, <span class="st">"ieu-a-7"</span>, <span class="kw">proxies</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="co">#&gt; Extracting data for 81 SNP(s) from 1 GWAS(s)</span>

<span class="co"># Harmonise the exposure and outcome data</span>
<span class="no">dat1</span> <span class="kw">&lt;-</span> <span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/harmonise_data.html">harmonise_data</a></span>(<span class="no">expd1</span>, <span class="no">outd1</span>)
<span class="co">#&gt; Harmonising LDL cholesterol || id:ieu-a-300 (ieu-a-300) and Coronary heart disease || id:ieu-a-7 (ieu-a-7)</span>
<span class="co">#&gt; Removing the following SNPs for being palindromic with intermediate allele frequencies:</span>
<span class="co">#&gt; rs2954029, rs964184</span>

<span class="co"># Perform MR</span>
<span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span>(<span class="no">dat1</span>)
<span class="co">#&gt; Analysing 'ieu-a-300' on 'ieu-a-7'</span>
<span class="co">#&gt;   id.exposure id.outcome                              outcome</span>
<span class="co">#&gt; 1   ieu-a-300    ieu-a-7 Coronary heart disease || id:ieu-a-7</span>
<span class="co">#&gt; 2   ieu-a-300    ieu-a-7 Coronary heart disease || id:ieu-a-7</span>
<span class="co">#&gt; 3   ieu-a-300    ieu-a-7 Coronary heart disease || id:ieu-a-7</span>
<span class="co">#&gt; 4   ieu-a-300    ieu-a-7 Coronary heart disease || id:ieu-a-7</span>
<span class="co">#&gt; 5   ieu-a-300    ieu-a-7 Coronary heart disease || id:ieu-a-7</span>
<span class="co">#&gt;                          exposure                    method nsnp         b</span>
<span class="co">#&gt; 1 LDL cholesterol || id:ieu-a-300                  MR Egger   78 0.5067936</span>
<span class="co">#&gt; 2 LDL cholesterol || id:ieu-a-300           Weighted median   78 0.4014675</span>
<span class="co">#&gt; 3 LDL cholesterol || id:ieu-a-300 Inverse variance weighted   78 0.4141687</span>
<span class="co">#&gt; 4 LDL cholesterol || id:ieu-a-300               Simple mode   78 0.4774092</span>
<span class="co">#&gt; 5 LDL cholesterol || id:ieu-a-300             Weighted mode   78 0.5328338</span>
<span class="co">#&gt;           se         pval</span>
<span class="co">#&gt; 1 0.07837328 8.725962e-09</span>
<span class="co">#&gt; 2 0.04575896 1.731524e-18</span>
<span class="co">#&gt; 3 0.05096262 4.403664e-16</span>
<span class="co">#&gt; 4 0.07618214 1.958799e-08</span>
<span class="co">#&gt; 5 0.07596319 7.807830e-10</span></pre></body></html></div>
</div>
<div id="using-gwas-vcf-files" class="section level2">
<h2 class="hasAnchor">
<a href="#using-gwas-vcf-files" class="anchor"></a>Using GWAS VCF files</h2>
<p>Let’s do the same with the vcf files (and the indexes). Download from here:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">wget</span> https://gwas.mrcieu.ac.uk/files/ieu-a-300/ieu-a-300.vcf.gz</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">wget</span> https://gwas.mrcieu.ac.uk/files/ieu-a-300/ieu-a-300.vcf.gz.tbi</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="fu">wget</span> https://gwas.mrcieu.ac.uk/files/ieu-a-7/ieu-a-7.vcf.gz</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="fu">wget</span> https://gwas.mrcieu.ac.uk/files/ieu-a-7/ieu-a-7.vcf.gz.tbi</span></code></pre></div>
<p>First get the tophits for LDL cholesterol</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/set_bcftools.html">set_bcftools</a></span>()
<span class="co">#&gt; Loading required namespace: genetics.binaRies</span>
<span class="co">#&gt; Path not provided, using binaries in the explodecomputer/genetics.binaRies package</span>
<span class="no">expd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/query_gwas.html">query_gwas</a></span>(<span class="st">"ieu-a-300.vcf.gz"</span>, <span class="kw">chrompos</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">expd1</span>$<span class="no">chr.exposure</span>, <span class="st">":"</span>, <span class="no">expd1</span>$<span class="no">pos.exposure</span>))</pre></body></html></div>
<p>Convert to TwoSampleMR format:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">expd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasglue</span><span class="kw ns">::</span><span class="fu"><a href="../reference/gwasvcf_to_TwoSampleMR.html">gwasvcf_to_TwoSampleMR</a></span>(<span class="no">expd2</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"exposure"</span>)</pre></body></html></div>
<p>Extract those SNPs from the outcome vcf file and convert to TwoSampleMR format</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">outd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/query_gwas.html">query_gwas</a></span>(<span class="st">"ieu-a-7.vcf.gz"</span>, <span class="kw">chrompos</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">expd1</span>$<span class="no">chr.exposure</span>, <span class="st">":"</span>, <span class="no">expd1</span>$<span class="no">pos.exposure</span>))
<span class="no">outd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasglue</span><span class="kw ns">::</span><span class="fu"><a href="../reference/gwasvcf_to_TwoSampleMR.html">gwasvcf_to_TwoSampleMR</a></span>(<span class="no">outd2</span>, <span class="st">"outcome"</span>)</pre></body></html></div>
<p>Proceed with harmonising and performing MR</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">dat2</span> <span class="kw">&lt;-</span> <span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/harmonise_data.html">harmonise_data</a></span>(<span class="no">expd2</span>, <span class="no">outd2</span>)
<span class="co">#&gt; Harmonising IEU-a-300 (PBNYCc) and IEU-a-7 (b4ucHT)</span>
<span class="co">#&gt; Removing the following SNPs for being palindromic with intermediate allele frequencies:</span>
<span class="co">#&gt; rs2954029, rs3184504, rs964184</span>
<span class="kw pkg">TwoSampleMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/TwoSampleMR/man/mr.html">mr</a></span>(<span class="no">dat2</span>)
<span class="co">#&gt; Analysing 'PBNYCc' on 'b4ucHT'</span>
<span class="co">#&gt;   id.exposure id.outcome outcome  exposure                    method nsnp</span>
<span class="co">#&gt; 1      PBNYCc     b4ucHT IEU-a-7 IEU-a-300                  MR Egger   77</span>
<span class="co">#&gt; 2      PBNYCc     b4ucHT IEU-a-7 IEU-a-300           Weighted median   77</span>
<span class="co">#&gt; 3      PBNYCc     b4ucHT IEU-a-7 IEU-a-300 Inverse variance weighted   77</span>
<span class="co">#&gt; 4      PBNYCc     b4ucHT IEU-a-7 IEU-a-300               Simple mode   77</span>
<span class="co">#&gt; 5      PBNYCc     b4ucHT IEU-a-7 IEU-a-300             Weighted mode   77</span>
<span class="co">#&gt;           b         se         pval</span>
<span class="co">#&gt; 1 0.4731258 0.07572869 2.299636e-08</span>
<span class="co">#&gt; 2 0.3976896 0.04383111 1.155429e-19</span>
<span class="co">#&gt; 3 0.4142950 0.04880527 2.088973e-17</span>
<span class="co">#&gt; 4 0.4476460 0.08095132 4.342522e-07</span>
<span class="co">#&gt; 5 0.5299800 0.07034075 8.526606e-11</span></pre></body></html></div>
</div>
<div id="other-options" class="section level2">
<h2 class="hasAnchor">
<a href="#other-options" class="anchor"></a>Other options</h2>
<div id="clumping-vcf-files" class="section level3">
<h3 class="hasAnchor">
<a href="#clumping-vcf-files" class="anchor"></a>Clumping vcf files</h3>
<p>If we want to extract top hits based on a threshold and clump locally. First download the LD reference dataset:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">wget</span> http://fileserve.mrcieu.ac.uk/ld/data_maf0.01_rs_ref.tgz</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="fu">tar</span> xzvf data_maf0.01_rs_ref.tgz</span></code></pre></div>
<p>Now extract the top hits based on a p-value threshold</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/set_bcftools.html">set_bcftools</a></span>()
<span class="co">#&gt; Path not provided, using binaries in the explodecomputer/genetics.binaRies package</span>
<span class="no">expd3</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/query_gwas.html">query_gwas</a></span>(<span class="st">"ieu-a-300.vcf.gz"</span>, <span class="kw">pval</span><span class="kw">=</span><span class="fl">5e-8</span>)
<span class="no">expd3</span>
<span class="co">#&gt; class: CollapsedVCF </span>
<span class="co">#&gt; dim: 3051 1 </span>
<span class="co">#&gt; rowRanges(vcf):</span>
<span class="co">#&gt;   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span>
<span class="co">#&gt; info(vcf):</span>
<span class="co">#&gt;   DataFrame with 3 columns: AF, AC, AN</span>
<span class="co">#&gt; info(header(vcf)):</span>
<span class="co">#&gt;       Number Type    Description                                </span>
<span class="co">#&gt;    AF A      Float   Allele Frequency                           </span>
<span class="co">#&gt;    AC A      Integer Allele count in genotypes                  </span>
<span class="co">#&gt;    AN 1      Integer Total number of alleles in called genotypes</span>
<span class="co">#&gt; geno(vcf):</span>
<span class="co">#&gt;   SimpleList of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span>
<span class="co">#&gt; geno(header(vcf)):</span>
<span class="co">#&gt;       Number Type   Description                                                </span>
<span class="co">#&gt;    ES A      Float  Effect size estimate relative to the alternative allele    </span>
<span class="co">#&gt;    SE A      Float  Standard error of effect size estimate                     </span>
<span class="co">#&gt;    LP A      Float  -log10 p-value for effect estimate                         </span>
<span class="co">#&gt;    AF A      Float  Alternate allele frequency in the association study        </span>
<span class="co">#&gt;    SS A      Float  Sample size used to estimate genetic effect                </span>
<span class="co">#&gt;    EZ A      Float  Z-score provided if it was used to derive the EFFECT and...</span>
<span class="co">#&gt;    SI A      Float  Accuracy score of summary data imputation                  </span>
<span class="co">#&gt;    NC A      Float  Number of cases used to estimate genetic effect            </span>
<span class="co">#&gt;    ID 1      String Study variant identifier</span></pre></body></html></div>
<p>Convert to TwoSampleMR format:</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">expd3</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasglue</span><span class="kw ns">::</span><span class="fu"><a href="../reference/gwasvcf_to_TwoSampleMR.html">gwasvcf_to_TwoSampleMR</a></span>(<span class="no">expd3</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"exposure"</span>)</pre></body></html></div>
<p>Get a list of SNPs to retain after clumping and subset the data</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">retain_snps</span> <span class="kw">&lt;-</span> <span class="no">expd3</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw">rsid</span><span class="kw">=</span><span class="no">SNP</span>, <span class="kw">pval</span><span class="kw">=</span><span class="no">pval.exposure</span>) <span class="kw">%&gt;%</span>
    <span class="kw pkg">ieugwasr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/ld_clump.html">ld_clump</a></span>(<span class="no">.</span>, <span class="kw">plink_bin</span><span class="kw">=</span><span class="kw pkg">genetics.binaRies</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/genetics.binaRies/man/get_plink_binary.html">get_plink_binary</a></span>(), <span class="kw">bfile</span><span class="kw">=</span><span class="st">"data_maf0.01_rs_ref"</span>) <span class="kw">%&gt;%</span>
    {<span class="no">.</span>$<span class="no">rsid</span>}
<span class="co">#&gt; Clumping IJQWqs, 3051 variants, using EUR population reference</span>
<span class="co">#&gt; Removing 2973 of 3051 variants due to LD with other variants or absence from LD reference panel</span>
<span class="no">expd3</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(<span class="no">expd3</span>, <span class="no">SNP</span> <span class="kw">%in%</span> <span class="no">retain_snps</span>)</pre></body></html></div>
</div>
<div id="extracting-outcome-data-with-ld-proxies" class="section level3">
<h3 class="hasAnchor">
<a href="#extracting-outcome-data-with-ld-proxies" class="anchor"></a>Extracting outcome data with LD proxies</h3>
<p>This only works if you extract on rsids at the moment, and is quite slow. But here it is:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/set_plink.html">set_plink</a></span>()
<span class="co">#&gt; Path not provided, using binaries in the explodecomputer/genetics.binaRies package</span>
<span class="no">outd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasvcf</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/query_gwas.html">query_gwas</a></span>(<span class="st">"ieu-a-7.vcf.gz"</span>, <span class="kw">rsid</span> <span class="kw">=</span> <span class="no">expd3</span>$<span class="no">SNP</span>, <span class="kw">proxies</span><span class="kw">=</span><span class="st">"yes"</span>, <span class="kw">bfile</span><span class="kw">=</span><span class="st">"data_maf0.01_rs_ref"</span>)
<span class="co">#&gt; Initial search...</span>
<span class="co">#&gt; Extracted 77 out of 78 rsids</span>
<span class="co">#&gt; Searching for proxies for 1 rsids</span>
<span class="co">#&gt; Determining searchspace...</span>
<span class="co">#&gt; Proxy lookup...</span>
<span class="co">#&gt; Finding proxies...</span>
<span class="co">#&gt; Taking input= as a system command ('gunzip -c /var/folders/8t/4mzsfqzd5773rsc11w4q5fhw0000gp/T//Rtmpg2bC85/filed244593ebd3e.targets.ld.gz') and a variable has been used in the expression passed to `input=`. Please use fread(cmd=...). There is a security concern if you are creating an app, and the app could have a malicious user, and the app is not running in a secure environment; e.g. the app is running as root. Please read item 5 in the NEWS file for v1.11.6 for more information and for the option to suppress this message.</span>
<span class="co">#&gt; Found 426 proxies</span>
<span class="co">#&gt; Extrating proxies...</span>
<span class="co">#&gt; Identified proxies for 1 of 1 rsids</span>
<span class="co">#&gt; Aligning...</span>
<span class="no">outd2</span> <span class="kw">&lt;-</span> <span class="kw pkg">gwasglue</span><span class="kw ns">::</span><span class="fu"><a href="../reference/gwasvcf_to_TwoSampleMR.html">gwasvcf_to_TwoSampleMR</a></span>(<span class="no">outd2</span>, <span class="st">"outcome"</span>)</pre></body></html></div>
</div>
</div>
<div id="further-mr-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#further-mr-methods" class="anchor"></a>Further MR methods</h2>
<p>A number of other MR methods are available. The current framework for using them is to:</p>
<ol style="list-style-type: decimal">
<li>Convert your data to TwoSampleMR format</li>
<li>Use the TwoSampleMR package to convert to other formats</li>
</ol>
<p>Examples of formats that you can convert to within TwoSampleMR:</p>
<ul>
<li><a href="https://github.com/rondolab/MR-PRESSO">MR-PRESSO</a></li>
<li><a href="https://github.com/WSpiller/RadialMR">RadialMR</a></li>
<li><a href="https://github.com/gqi/MRMix/">MRMix</a></li>
<li><a href="https://cran.r-project.org/web/packages/MendelianRandomization/index.html">MendelianRandomization</a></li>
<li><a href="https://cran.r-project.org/web/packages/mr.raps/">MR-RAPS</a></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
